File: leetcode_ly/java/Array/JumpGame.java
public class Solution {
    public boolean canJump(int[] nums) {
        int[] canWalk = new int[nums.length];
        canWalk[0] = nums[0]; 
        for(int i = 1; i < nums.length; i++){
            canWalk[i] = Math.max(canWalk[i - 1],nums[i - 1]) - 1;
            if(canWalk[i] < 0) return false;
        }
        return canWalk[nums.length - 1] >= 0;
    }
}

File: leetcode_ly/c++/DynamicProgramming/JumpGame.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int len = nums.size();
        if(len <= 1) return true;
        if(nums[0] >= len - 1) return true;
        
        int* canWalk = new int[len];
        canWalk[0] = nums[0];
        for(int i = 1; i < len; i++){
            int max = (canWalk[i - 1] > nums[i - 1]) ? canWalk[i - 1] : nums[i - 1];
            canWalk[i] = max - 1;
            if(canWalk[i] < 0) return false;
        }
        return canWalk[len - 1] >= 0;
    }
};